Add cancellation functionality to progress

gselzer opened the pull request:
# Description
<!-- What does this pull request (PR) do? Why is it necessary? -->
<!-- Tell us about your new feature, improvement, or fix! -->
<!-- If your change includes user interface changes, please add an image, or an animation "An image is worth a thousand words!" -->
<!-- You can use https://www.cockos.com/licecap/ or similar to create animations -->

This PR provides the ability for users to cancel tasks utilizing the `progress` mechanism. Therefore, if you write something like

```python
for i in progress(range(10)):
        do_something_time_consuming()
```
and the user decides to cancel on the fifth iteration of the `for` loop, the remaining 5 iterations will **not** be run. The user can assert cancellation with the field `progress.is_canceled()`, and can provide cancellation error-handling by passing a `Callable` to the `progress` object as the `cancel_callback` parameter - this can enable context destruction in the case of cancellation.



Open questions:
* What should the behavior be when there are inevitably lines beyond the `for` loop which expect the loop to have completed as expected? I see two possibilities:
  * We raise an exception of some sort within `progress.cancel`.
  * We require the remaining code to be conditional on `progress.is_cancel` returning `False

  I prefer the former, as it could be tough to check that if we don't retain a reference to progress. Retaining that reference would increase verbosity.
* Should we provide any kwargs to the cancellation callback? The cancellation callback might want to know e.g. how far computation got before it was cancelled. But I don't know what you'd want. We could also wait until someone requests that functionality to implement it.

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->

## Type of change
<!-- Please delete options that are not relevant. -->
- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [X] test suite for my feature covers cases x, y, and z
- [X] all tests pass with my change

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

gselzer committed:
Add cancellation functionality to progress

gselzer committed:
Fix Wrapped superclass

gselzer committed:
Use translation for strings

gselzer committed:
Rewrite cancellable progress as a new class

Czaki commented on napari/_qt/widgets/qt_progress_bar.py:
```diff
@@ -44,6 +46,11 @@ def __init__(
+            self.cancel_button = QPushButton(trans._('Cancel'))
+            self.cancel_button.clicked.connect(self._cancel)
+            pbar_layout.addWidget(self.cancel_button)
+
```
Conditional adding attribute is not the best practice. Maybe `self.cacnel_button.setVisible(isinstance(prog, cancelable_progress))`

gselzer replied:
Added it in cb34bc93

Czaki commented on napari/utils/progress.py:
```diff
@@ -150,3 +150,42 @@ def progrange(*args, **kwargs):
+                    raise StopIteration
+                return next(plain_iterator)
+
+        return Wrapped()
```
Maybe takewhile may be used? https://docs.python.org/3/library/itertools.html#itertools.takewhile

gselzer replied:
Great idea, fixed with 6b4ddab

Czaki reviewed (commented):
did I correctly found "clunky" places

gselzer committed:
Use itertools.takeWhile

gselzer committed:
Add setVisible conditional, and test Widget

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

DragaDoncila commented on napari/utils/progress.py:
```diff
@@ -150,3 +151,30 @@ def progrange(*args, **kwargs):
+        super().__init__(iterable, desc, total, nest_under, *args, **kwargs)
+
+    def __iter__(self) -> Iterator:
+        return takewhile(lambda _: not self.is_canceled, self.iterable)
```
This leaves us unable to see progress bar updates in the GUI, and the bar doesn't disappear after cancellation. I think it should be:

```suggestion
        return takewhile(lambda _: not self.is_canceled, super().__iter__())
```

DragaDoncila reviewed (changes requested):
@gselzer thanks for opening this! I've had a chance to have a look now and I like the implementation. I agree with Czaki's original suggestion that it should be its own class.

I left a comment on the `takewhile` line - we need to pass it the super-class' iterator so that it can emit the correct events for updating the GUI - currently the GUI progress bar doesn't do anything.

It would be great to add a `cancelable_progress` example to the `progress_minimal.py` example e.g.

```python
def cancel_callback():
    print("Canceling") # would be nice if this was something more useful but.. not sure what it could be
    
def cancelable_iterable():
    for _ in cancelable_progress(range(100), cancel_callback=cancel_callback):
        np.random.rand(128, 128, 128).mean(0)
        
cancel_iter_btn = QPushButton("Cancelable Iterable")
cancel_iter_btn.clicked.connect(cancelable_iterable)
button_layout.addWidget(cancel_iter_btn)
```

Otherwise looking good! Next step would then be to add the cancel functionality to the generator threadworkers in `_qt/threading/create_worker`, but that's a follow-up PR imo.

gselzer committed:
use super().__iter__()

When switching to takewhile, I accidentally used the wrong iterable!

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

gselzer committed:
Add example to progress_bar_minimal.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

gselzer committed:
Cancel progress within take_while

This prevents any asynchronous funny business, where cancel_callback is
called by the GUI thread in the middle of an iteration

gselzer committed:
Add documentation for cancelable_progress class

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

DragaDoncila reviewed (approved):
Thank you for making those changes @gselzer! Agreed, checking for cancelation in the `takewhile` makes more sense and is less likely to lead to chaos. This looks good to me, excited to get it in! @Czaki did you want to take a final pass?

Czaki commented on napari/utils/progress.py:
```diff
@@ -150,3 +151,59 @@ def progrange(*args, **kwargs):
+
+        return takewhile(is_canceled, super().__iter__())
+
+    def cancel(self):
```
This method should have a docstring to be visible in API documentation.

gselzer replied:
Docstring added in fb518dc8

gselzer committed:
Add docstring for cancelable_progress.cancel

gselzer committed:
Add generator.close() call when necessary

Czaki commented on napari/utils/_tests/test_progress.py:
```diff
@@ -93,3 +93,59 @@ def test_progrange():
+                yield i
+                i += 1
+        finally:
+            global closed
```
```suggestion
            nonlocal closed
```

Closed is not a module-level variable.

Czaki reviewed (commented):
Ok. There is a bug in tqdm. I will create PR.

Czaki commented on napari/utils/progress.py:
```diff
@@ -150,3 +151,68 @@ def progrange(*args, **kwargs):
+                    self.cancel_callback()
+                    # Perform additional cleanup for generators
+                    if isinstance(itr, Generator):
+                        itr.close()
```
```suggestion
                    if isinstance(self.iterable, Generator):
                        self.iterable.close()
```

The `tqdm` exposes the inner iterable object on attribute `iterable`.

Czaki reviewed (commented):
After local testing, it looks like these two changes should be enough to get the code working.

gselzer committed:
Close self.iterable(), use nonlocal in test

Czaki committed:
Merge branch 'main' into progress_cancellation
